File: frontend/src/types/pipeline.ts
/**
 * Pipeline type definitions
 */

/**
 * 步骤分类
 */
export type StepCategory = 'preprocess' | 'analysis';

/**
 * 预处理方法
 */
export type PreprocessMethod =
  | 'none'
  | 'linear_detrend'
  | 'polynomial_detrend'
  | 'highpass_filter'
  | 'bandpass_filter'
  | 'lowpass_filter'
  | 'notch_filter'
  | 'baseline_correction';

/**
 * 分析方法
 */
export type AnalysisMethod = 'anomaly_detection' | 'artifact_detection' | 'band_power';

/**
 * 流水线步骤
 */
export interface PipelineStep {
  id: string;
  category: StepCategory;
  method: PreprocessMethod | AnalysisMethod;
  parameters: Record<string, number>;
  enabled: boolean;
  order: number;
}

/**
 * 步骤执行日志
 */
export interface StepLog {
  step_id: string;
  status: 'success' | 'error' | 'skipped';
  duration_ms: number;
  message: string;
}

/**
 * 处理后波形数据
 */
export interface ProcessedWaveformData {
  data: number[][];
  times: number[];
  channels: string[];
  sfreq: number;
  n_samples: number;
  start_time: number;
  duration: number;
}

/**
 * 流水线 API 请求
 */
export interface PipelineRequest {
  steps: PipelineStep[];
  start: number;
  duration: number;
  channels?: string[] | null;
}

/**
 * 流水线 API 响应
 */
export interface PipelineResponse {
  processed_waveform: ProcessedWaveformData | null;
  annotations: PipelineAnnotation[];
  step_logs: StepLog[];
  total_processing_time_ms: number;
}

/**
 * 流水线产生的标注
 */
export interface PipelineAnnotation {
  annotation_type: string;
  source: 'pipeline';
  channel: string | null;
  start_time: number;
  end_time: number;
  label: string;
  color: string;
  severity: number;
  confidence: number;
  metadata: Record<string, unknown>;
}

/**
 * 步骤定义（用于 UI 展示）
 */
export interface StepDefinition {
  name: string;
  description: string;
  category: StepCategory;
  parameters?: Record<string, {
    type: 'int' | 'float';
    default: number;
    min: number;
    max: number;
    description: string;
  }>;
}

/**
 * 预处理步骤定义
 */
export const PREPROCESS_STEP_DEFINITIONS: Record<PreprocessMethod, StepDefinition> = {
  none: {
    name: '无预处理',
    description: '保持原始信号',
    category: 'preprocess',
  },
  linear_detrend: {
    name: '线性去漂移',
    description: '去除线性趋势，适用于线性漂移',
    category: 'preprocess',
  },
  polynomial_detrend: {
    name: '多项式去漂移',
    description: '去除多项式趋势，适用于复杂漂移',
    category: 'preprocess',
    parameters: {
      order: { type: 'int', default: 2, min: 1, max: 5, description: '多项式阶数' },
    },
  },
  highpass_filter: {
    name: '高通滤波',
    description: '去除低频成分，适用于低频漂移',
    category: 'preprocess',
    parameters: {
      cutoff: { type: 'float', default: 0.5, min: 0.1, max: 2.0, description: '截止频率 (Hz)' },
    },
  },
  bandpass_filter: {
    name: '带通滤波',
    description: '保留特定频率范围',
    category: 'preprocess',
    parameters: {
      lowcut: { type: 'float', default: 0.5, min: 0.1, max: 10.0, description: '低截止频率 (Hz)' },
      highcut: { type: 'float', default: 50.0, min: 10.0, max: 100.0, description: '高截止频率 (Hz)' },
    },
  },
  lowpass_filter: {
    name: '低通滤波',
    description: '去除高频噪声',
    category: 'preprocess',
    parameters: {
      cutoff: { type: 'float', default: 50.0, min: 10.0, max: 100.0, description: '截止频率 (Hz)' },
    },
  },
  notch_filter: {
    name: '陷波滤波',
    description: '去除工频干扰 (50Hz/60Hz)',
    category: 'preprocess',
    parameters: {
      freq: { type: 'float', default: 50.0, min: 49.0, max: 61.0, description: '陷波频率 (Hz)' },
      quality: { type: 'float', default: 30.0, min: 10.0, max: 50.0, description: '品质因数' },
    },
  },
  baseline_correction: {
    name: '基线校正',
    description: '使用移动平均去除基线',
    category: 'preprocess',
  },
};

/**
 * 分析步骤定义
 */
export const ANALYSIS_STEP_DEFINITIONS: Record<AnalysisMethod, StepDefinition> = {
  anomaly_detection: {
    name: '异常检测',
    description: '检测棘波、尖波等异常波形',
    category: 'analysis',
    parameters: {
      sensitivity: { type: 'float', default: 1.0, min: 0.5, max: 1.5, description: '检测灵敏度' },
    },
  },
  artifact_detection: {
    name: '伪迹检测',
    description: '检测眼电、肌电等伪迹',
    category: 'analysis',
  },
  band_power: {
    name: '频带功率分析',
    description: '分析各频带功率分布',
    category: 'analysis',
  },
};

/**
 * 默认步骤（用于初始化）
 */
export const DEFAULT_STEPS: PipelineStep[] = [
  {
    id: 'step-1',
    category: 'preprocess',
    method: 'bandpass_filter',
    parameters: { lowcut: 0.5, highcut: 50.0 },
    enabled: true,
    order: 0,
  },
  {
    id: 'step-2',
    category: 'analysis',
    method: 'anomaly_detection',
    parameters: { sensitivity: 1.0 },
    enabled: true,
    order: 1,
  },
];

File: backend/app/api/routes/pipeline.py
"""
Pipeline API endpoint - Sequential EEG processing pipeline
"""

import asyncio
import numpy as np
from fastapi import APIRouter, HTTPException
from pydantic import BaseModel
from typing import List, Dict, Any, Optional
from app.services.pipeline_service import PipelineService
from app.services.file_manager import get_file_path
import logging

logger = logging.getLogger(__name__)

router = APIRouter()


class PipelineStepRequest(BaseModel):
    id: str
    category: str  # "preprocess" | "analysis"
    method: str
    parameters: Optional[Dict[str, Any]] = None
    enabled: bool = True


class PipelineRequest(BaseModel):
    steps: List[PipelineStepRequest]
    start: float
    duration: float
    channels: Optional[List[str]] = None


class StepLog(BaseModel):
    step_id: str
    status: str  # "success" | "error" | "skipped"
    duration_ms: int
    message: str


class ProcessedWaveformData(BaseModel):
    data: List[List[float]]
    times: List[float]
    channels: List[str]
    sfreq: float
    n_samples: int
    start_time: float
    duration: float


class PipelineResponse(BaseModel):
    processed_waveform: Optional[ProcessedWaveformData] = None
    annotations: List[Dict[str, Any]] = []
    step_logs: List[StepLog] = []
    total_processing_time_ms: int = 0


@router.post("/pipeline/{file_id}", response_model=PipelineResponse)
async def run_pipeline(file_id: str, request: PipelineRequest):
    """
    执行处理流水线：串行预处理 → 分析 → 返回处理后波形和标注
    """
    try:
        file_path = get_file_path(file_id)

        steps = [s.model_dump() for s in request.steps]

        def _run():
            service = PipelineService()
            return service.run_pipeline(
                file_path=file_path,
                steps=steps,
                start=request.start,
                duration=request.duration,
                channels=request.channels,
            )

        result = await asyncio.to_thread(_run)

        processed_waveform = None
        if result["processed_data"] is not None:
            data = result["processed_data"]
            if isinstance(data, np.ndarray):
                data = data.tolist()
            processed_waveform = ProcessedWaveformData(
                data=data,
                times=result["times"],
                channels=result["channels"],
                sfreq=result["sfreq"],
                n_samples=result["n_samples"],
                start_time=result["start_time"],
                duration=result["duration"],
            )

        step_logs = [
            StepLog(**log) for log in result.get("step_logs", [])
        ]

        logger.info(
            f"Pipeline completed for {file_id}: "
            f"{len(result.get('annotations', []))} annotations, "
            f"{result.get('total_processing_time_ms', 0)}ms"
        )

        return PipelineResponse(
            processed_waveform=processed_waveform,
            annotations=result.get("annotations", []),
            step_logs=step_logs,
            total_processing_time_ms=result.get("total_processing_time_ms", 0),
        )

    except FileNotFoundError:
        logger.error(f"File not found: {file_id}")
        raise HTTPException(status_code=404, detail=f"文件未找到: {file_id}")
    except Exception as e:
        logger.error(f"Pipeline error: {e}")
        raise HTTPException(status_code=500, detail=f"流水线执行失败: {str(e)}")
